### PR TITLE
Docs, part 1

### PR DESCRIPTION
**Pull request by arogachev**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | #358 

- Add custom Psalm type `WhenType` and apply it across all project.
- Rename `SerializableRuleInterface` to `RuleWithOptionsInterface` because technically `RuleInterface` is serializable too but with name only.
- Add docs for `RuleInterface` and `RuleWithOptionsInterface`.

### CODE REVIEW

File: src/RuleWithOptionsInterface.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+     *
+     * ```php
+     * [
+     *     'property1' => $this->>property1,
+     *     'property2' => $this->>property2,
+     *     // ...
+     * ];
+     * ```
+     *
+     * Sometimes it's helpful to include dynamic options as well:
```
Comment (vjik) [inline src/RuleWithOptionsInterface.php:34]: What is dynamic options?

File: src/RuleWithOptionsInterface.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+     * ```php
+     * 'message' => [ // Array is used with no parameters too.
+     *     'template => 'A message without parameters.',
+     *     'parameters' => [], // Explicitly specified even for empty parameters.
+     * ],
+     * 'anotherMessage' => [
+     *     'template' => 'A message with {parameter}.', // Handler dependent parameters
+     *     'parameters' => ['property4' => $this->property4], // Property dependent parameters
+     * ],
+     * 'wrongMessage' => 'Wrong message.' // This is possible, but wrong. Use example above for consistent structure.
```
Comment (vjik) [inline src/RuleWithOptionsInterface.php:55]: `This is possible` — suggest remove this. By our agreements messages in options always must have structure:

```
'message' => ['template' => ..., 'parameters' => [...]]
```

File: src/RuleWithOptionsInterface.php (hunk 1 above)
Comment (arogachev) [reply src/RuleWithOptionsInterface.php:34]: For example:

https://github.com/yiisoft/validator/blob/f4e4de303fb05877af5aa6716d04dce579939cca/src/Rule/Compare.php#L169-L172

https://github.com/yiisoft/validator/blob/f4e4de303fb05877af5aa6716d04dce579939cca/src/Rule/Compare.php#L138-L148

The idea was to show that not only properties can be included as options.

File: src/RuleWithOptionsInterface.php (hunk 2 above)
Comment (arogachev) [reply src/RuleWithOptionsInterface.php]: Fixed. Just in case, you mean remove the comment part, right? Or remove this example completely?

Comment (vjik) [reply src/RuleWithOptionsInterface.php]: Now is fine!

File: src/RuleWithOptionsInterface.php (hunk 1 above)
Comment (vjik) [reply src/RuleWithOptionsInterface.php:34]: May be completely remove this block? It's base PHP availabilities.

Comment (arogachev) [reply src/RuleWithOptionsInterface.php]: OK. Removed it completely.

**Review by vjik — APPROVED** [review]

**Pull request merged**